server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java

```diff
@@ -70,6 +74,87 @@ protected boolean addMockInternalEngine() {
         return false;
     }
 
+    public void testPrimaryStopped_ReplicaPromoted() throws Exception {
+        final String nodeA = internalCluster().startNode();
+        final String nodeB = internalCluster().startNode();
+        createIndex(INDEX_NAME);
+        ensureGreen(INDEX_NAME);
+
+        client().prepareIndex(INDEX_NAME).setId("1").setSource("foo", "bar").setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).get();
+        refresh(INDEX_NAME);
+
+        waitForReplicaUpdate();
+        assertHitCount(client(nodeA).prepareSearch(INDEX_NAME).setSize(0).setPreference("_only_local").get(), 1);
+        assertHitCount(client(nodeB).prepareSearch(INDEX_NAME).setSize(0).setPreference("_only_local").get(), 1);
+
+        final DiscoveryNode primaryDiscoveryNode = getNodeContainingPrimaryShard();
+        final String primaryNodeName = primaryDiscoveryNode.getName();
+        final String replicaNodeName = nodeA.equals(primaryNodeName) ? nodeB : nodeA;
+
+        // index another doc but don't refresh, we will ensure this is searchable once replica is promoted.
+        client().prepareIndex(INDEX_NAME).setId("2").setSource("bar", "baz").setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).get();
+
+        // stop the primary node - we only have one shard on here.
+        internalCluster().stopRandomNode(InternalTestCluster.nameFilter(primaryNodeName));
+        ensureYellowAndNoInitializingShards(INDEX_NAME);
+
+        final ShardRouting replicaShardRouting = getShardRoutingForNodeName(replicaNodeName);
+        assertNotNull(replicaShardRouting);
+        assertTrue(replicaShardRouting + " should be promoted as a primary", replicaShardRouting.primary());
+        assertHitCount(client(replicaNodeName).prepareSearch(INDEX_NAME).setSize(0).setPreference("_only_local").get(), 2);
+
+        // assert we can index into the new primary.
+        client().prepareIndex(INDEX_NAME).setId("3").setSource("bar", "baz").setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).get();
+        assertHitCount(client(replicaNodeName).prepareSearch(INDEX_NAME).setSize(0).setPreference("_only_local").get(), 3);
+
+        // start another node, index another doc and replicate.
+        String nodeC = internalCluster().startNode();
+        ensureGreen(INDEX_NAME);
+        client().prepareIndex(INDEX_NAME).setId("4").setSource("baz", "baz").get();
+        refresh(INDEX_NAME);
+        waitForReplicaUpdate();
+        assertHitCount(client(nodeC).prepareSearch(INDEX_NAME).setSize(0).setPreference("_only_local").get(), 4);
+        assertHitCount(client(replicaNodeName).prepareSearch(INDEX_NAME).setSize(0).setPreference("_only_local").get(), 4);
+        assertSegmentStats(REPLICA_COUNT);
+    }
+
+    private DiscoveryNode getNodeContainingPrimaryShard() {
```

Comment:
In tests, a index can be created only with one node which means primary will be assigned on that node (replica will be unassinged obviously) and then start another node, which will assign the replica shard on this second node.

Reply:
I've updated the test to follow the new sequence of startup steps, but I've left this for validation so I can assert the replica was promoted correctly.